src/extra/stream/client.test.ts

```diff
@@ -1,22 +1,94 @@
 // @vitest-environment happy-dom
 
-import { describe, expect, test } from 'vitest';
+import { vi, beforeEach, afterEach, describe, expect, test } from 'vitest';
 import { RobotClient } from '../../robot';
 import { events } from '../../events';
+import { StreamServiceClient } from '../../gen/proto/stream/v1/stream_pb_service';
 import { StreamClient } from './client';
 
+let robotClient: RobotClient;
+let streamClient: StreamClient;
+
 describe('StreamClient', () => {
+  beforeEach(() => {
+    vi.useFakeTimers();
+
+    vi.mock('./robot/client');
+
+    robotClient = new RobotClient('fakehost');
+    vi.mock('./gen/proto/stream/v1/stream_pb_service');
+
+    streamClient = new StreamClient(robotClient);
+  });
+
+  afterEach(() => {
+    vi.clearAllMocks();
+    vi.useRealTimers();
+  });
+
   test('webrtc track will cause the client to emit an event', () =>
     new Promise<void>((done) => {
-      const host = 'fakeServiceHost';
-      const client = new RobotClient(host);
-      const streamClient = new StreamClient(client);
-
       streamClient.on('track', (data) => {
         expect((data as { mock: true }).mock).eq(true);
         done();
       });
 
       events.emit('track', { mock: true });
     }));
+
+  test('getStream creates and returns a new stream', async () => {
+    const fakeCamName = 'fakecam';
+    const fakeStream = { id: fakeCamName };
+    StreamServiceClient.prototype.addStream = vi
+      .fn()
+      .mockImplementation((_req, _md, cb) => {
+        cb(null, {});
+        streamClient.emit('track', { streams: [fakeStream] });
```

Comment:
Should this be wrapped in an `if` that checks the arguments to ensure `addStream` is called correctly? This `mockImplementation` is unconditional, so `getStream` could call `addStream` incorrectly and this test would still pass

This question applies to all subsequent tests that fake out `addStream`. #92 shows an alternative approach that using a [library I wrote](https://mike.cousins.io/articles/2023-06-30-better-stubs/) that prevents unconditional mocking by design

Reply:
ah good call, will update